espnet/nets/pytorch_backend/transducer/rnn_decoder.py

```diff
@@ -0,0 +1,309 @@
+"""RNN-Transducer implementation for training and decoding."""
+
+import torch
+
+from espnet.nets.pytorch_backend.nets_utils import get_activation
+from espnet.nets.pytorch_backend.nets_utils import to_device
+
+from espnet.nets.transducer_decoder_interface import TransducerDecoderInterface
+
+
+class DecoderRNNT(TransducerDecoderInterface, torch.nn.Module):
+    """RNN-T Decoder module.
+
+    Args:
+        eprojs (int): # encoder projection units
+        odim (int): dimension of outputs
+        dtype (str): gru or lstm
+        dlayers (int): # prediction layers
+        dunits (int): # prediction units
+        blank (int): blank symbol id
+        embed_dim (init): dimension of embeddings
+        joint_dim (int): dimension of joint space
+        joint_activation_type (int): joint network activation
+        dropout (float): dropout rate
+        dropout_embed (float): embedding dropout rate
+
+    """
+
+    def __init__(
+        self,
+        eprojs,
+        odim,
+        dtype,
+        dlayers,
+        dunits,
+        blank,
+        embed_dim,
+        joint_dim,
+        joint_activation_type="tanh",
+        dropout=0.0,
+        dropout_embed=0.0,
+    ):
+        """Transducer initializer."""
+        super(DecoderRNNT, self).__init__()
+
+        self.embed = torch.nn.Embedding(odim, embed_dim, padding_idx=blank)
+        self.dropout_embed = torch.nn.Dropout(p=dropout_embed)
+
+        if dtype == "lstm":
+            dec_net = torch.nn.LSTMCell
+        else:
+            dec_net = torch.nn.GRUCell
+
+        self.decoder = torch.nn.ModuleList([dec_net(embed_dim, dunits)])
+        self.dropout_dec = torch.nn.ModuleList([torch.nn.Dropout(p=dropout)])
+
+        for _ in range(1, dlayers):
+            self.decoder += [dec_net(dunits, dunits)]
+            self.dropout_dec += [torch.nn.Dropout(p=dropout)]
+
+        self.lin_enc = torch.nn.Linear(eprojs, joint_dim)
+        self.lin_dec = torch.nn.Linear(dunits, joint_dim, bias=False)
+        self.lin_out = torch.nn.Linear(joint_dim, odim)
+
+        self.joint_activation = get_activation(joint_activation_type)
+
+        self.dlayers = dlayers
+        self.dunits = dunits
+        self.dtype = dtype
+        self.embed_dim = embed_dim
+        self.joint_dim = joint_dim
+        self.odim = odim
+
+        self.ignore_id = -1
+        self.blank = blank
+
+    def init_state(self, init_tensor):
+        """Initialize decoder states.
+
+        Args:
+            init_tensor (torch.Tensor): batch of input features
+                (B, emb_dim / dec_dim)
+
+        Returns:
+            (tuple): batch of decoder states
+                ([L x (B, dec_dim)], [L x (B, dec_dim)])
+
+        """
+        z_list = [
+            to_device(self, torch.zeros(init_tensor.size(0), self.dunits))
```

Comment:
Did you follow #2492 #2169 ?

`to_device(self, tensor)` can't be used for pytorch>=1.5.
We must solve this problem for training mode at least, so It's not better to use to_device in `forward()`. It's okay to use in decode function, I think.

Could you fix it at the next PR?

Reply:
I didn't, sorry! I'll fix that in another PR!